server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -352,19 +352,18 @@ private synchronized void onMergeFailure(Exception exc) {
                 circuitBreakerBytes = 0;
             }
             failure.compareAndSet(null, exc);
-            MergeTask task = runningTask.get();
-            runningTask.compareAndSet(task, null);
-            onPartialMergeFailure.accept(exc);
-            List<MergeTask> toCancels = new ArrayList<>();
+            final MergeTask task = runningTask.getAndSet(null);
+            final List<Releasable> toCancels = new ArrayList<>();
+            toCancels.add(() -> onPartialMergeFailure.accept(exc));
             if (task != null) {
-                toCancels.add(task);
+                toCancels.add(task::cancel);
+            }
+            for (MergeTask mergeTask : queue) {
```

Comment:
I was wondering if could clear the queue and iterate it one go with
```java
while (queue.isEmpty() == false){
      toCancels.add(queue.removeFirst()::cancel);
}
```

Reply:
Sure saves a line :) I suppose we can even just `pollFirst` in a loop :) Will do.